django/thunderstore/repository/api/experimental/serializers.py

```diff
@@ -95,3 +96,132 @@ class Meta:
             "has_nsfw_content",
             "categories",
         )
+
+
+class PackageUploadAuthorNameField(serializers.SlugRelatedField):
+    def __init__(self, *args, **kwargs):
+        kwargs["slug_field"] = "name"
+        super().__init__(*args, **kwargs)
+
+    def get_queryset(self):
+        return UploaderIdentity.objects.filter(
+            members__user=self.context["request"].user,
+        )
+
+
+class PackageUploadCategoriesField(serializers.RelatedField):
+    def get_queryset(self):
+        return PackageCategory.objects.filter(
+            community=self.context["request"].community,
+        )
+
+    def to_representation(self, value):
+        return [c.pk for c in value]
+
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            raise serializers.ValidationError("Not a list")
+
+        out = []
+        for category_pk in data:
+            category = self.get_queryset().filter(pk=category_pk).first()
+            if not category:
+                raise serializers.ValidationError(f"{category_pk} category not found")
+            out.append(category)
+        return out
+
+
+class PackageUploadSerializer(serializers.Serializer):
+    author_name = PackageUploadAuthorNameField()
+    categories = PackageUploadCategoriesField()
+    has_nsfw_content = serializers.BooleanField()
+    pk = serializers.IntegerField(read_only=True)
+    file = serializers.FileField(write_only=True)
+
+    def _create_form(self, data) -> PackageUploadForm:
+        request = self.context["request"]
+        return PackageUploadForm(
+            request.user,
+            data.get("author_name"),
+            request.community,
+            data={
+                "categories": data.get("categories"),
+                "has_nsfw_content": data.get("has_nsfw_content"),
+            },
+            files={"file": data.get("file")},
+        )
+
+    def validate(self, data):
+        form = self._create_form(data)
+        if not form.is_valid():
+            raise serializers.ValidationError("Invalid")
```

Comment:
You can get validation errors from the form itself, and raise those here. Validation errors also support a list of messages instead of just a single message.